src/lib/fhirValidator.js

```
@@ -0,0 +1,86 @@
+		if (this.conformance !== undefined) {
+			return Promise.resolve(this.conformance);
+		}
+
+		return request.get(config.fhirSchemaUrl)
+			.then((res) => {
+				const schema = JSON.parse(res.text);
+				Object.keys(schema.resources).forEach((key) => {
+					schema.resources[key].types = Object.assign({},
+						schema.resources[key].types, schema.types);
+				});
+				this.conformance = {
+					types: schema.types,
+					resources: schema.resources,
+				};
+				return this.conformance;
+			});
+	}
+
+	isValidResourceType(resourceType) {
+		if (resourceType.includes('.')) return new Promise((resolve, reject) => reject(false));
+		return this.getConformance()
+			.then(conform => Object.prototype.hasOwnProperty.call(conform.resources, resourceType));
+	}
+
+	validate(data) {
+		let result;
+		let	resource;
+		let	valid;
+
+		result = {
+			resource: null,
+			errors: [],
+		};
+
+		// Ensure that the data provided is a javascript object.
+		// If it is a string then attempt to parse it
+		try {
+			if (typeof data === 'string') resource = JSON.parse(data);
+			else if (typeof data === 'object') resource = data;
+			else {
+				result.errors.push(new ValidationError(
+					'Data provided must be a string or plain javascript object',
+				));
+				return Promise.reject(result);
+			}
+			result.resource = resource;
+		} catch (e) {
+			result.errors.push(new ValidationError('Invalid JSON', e));
+			return result;
+		}
+
+		return this.isValidResourceType(resource.resourceType)
+			.then((res) => {
+				if (res === true) {
+					const validate = ajv.compile(this.conformance.resources[resource.resourceType]);
+					validate(resource);
+					if (validate.errors) {
+						result.errors = result.errors.concat(validate.errors);
+					}
+				} else {
+					result.errors.push(new ValidationError('Not a valid resource type'));
+				}
+				return result;
+			})
+			.catch(() => {
+				result.errors.push(new ValidationError('Not a valid resource type'));
+				return result;
```

Why don't you better throw an error instead of pushing to results and return it?